src/modm/driver/motion/adns9800.hpp

```diff
@@ -9,95 +9,426 @@
+    }
+  }
+
+  void
+  setExposureMax(const PeriodType shutter_max) {
+    const PeriodType recover = shutter_config.shutter_max;
+    shutter_config.shutter_max = shutter_max;
+
+    if (shutter_config.assert()) {
+      // @todo waiting for #690 that brings 16bit Spi transfers
+      writeRegister(Register::Shutter_Max_Bound_Lower, shutter_config.shutter_max & 0xff);
+      writeRegister(Register::Shutter_Max_Bound_Upper, shutter_config.shutter_max >> 8);
+    } else {
+      shutter_config.shutter_max = recover;
+    }
+  }
+
+  /**
+   * Enable or disable the Angle Snapping function. When enabled, actual movement ranges from ±5° from X or Y-axis,
+   * it will be snapped to the closest axis. For example, if the sensor moves at ±2° from X-axis, the output motion
+   * data will be snapped to X-axis at 0°.
+   */
+  void
+  setSnapAngle(bool const value) {
+    const uint8_t preserved = readRegister(Register::Snap_Angle) & ~0x80;
+    writeRegister(Register::Snap_Angle, preserved | value ? 0x80 : 0x00);
+  }
+
+  bool
+  hasNewMotionData() {
+    return readRegister(Register::Motion) & Bit7;
+  }
+
+  /**
+   * Read the latest motion delta from the sensor and update data with it content
+   * Reading the motion delta also reloads the Motion interrupt (Pin7)
+   *
+   * @tparam      One of Data, DataObserve or DataAnalysis depending on your need for details
+   */
+  template<class D>
+  requires std::is_base_of_v<Data, D>
+  D
+  read() {
+    uint8_t buffer[D::length];
+
+    readTransaction([this, &buffer]() {
+      SpiMaster::transfer(static_cast<uint8_t>(Register::Motion_Burst));
+      modm::this_fiber::sleep_for(shutter_config.getOneFrameTime());
+      SpiMaster::transfer(nullptr, buffer, sizeof(buffer));
+    });
+
+    return D(buffer);
+  }
+
+  // @todo test captureFrame
+  void
+  captureFrame(std::span<uint8_t, FrameSize> buffer) {
+    readTransaction([&buffer]() {
+      writeRegister(Register::Frame_Capture, 0x93);
+      writeRegister(Register::Frame_Capture, 0xc5);
+      modm::this_fiber::poll(readRegister(Register::Motion) & Bit0); // wait for first pixel
+      SpiMaster::transfer(nullptr, buffer, FrameSize);
+    });
+  }
+
+private:
+  template<class Callable>
+  void
+  readTransaction(Callable&& closure) {
+    while(not ready_to_read.isExpired()) modm::this_fiber::yield();
+    while(not this->acquireMaster()) modm::this_fiber::yield();
+    Cs::reset();
+
+    closure();
+
+    ready_to_write.restart(20us);      // tSRW: [t]ime [S]pi between [R]rite and [W]rite
+    ready_to_read.restart(20us);       // tSRR: [t]ime [S]pi between [R]rite and [R]ead
+
+    modm::this_fiber::sleep_for(120ns); // tSCLK_NCS_read: Cs persistance after last SCLK for read
+    Cs::set();
+  }
+
+  template<class Callable>
+  void
+  writeTransaction(Callable&& closure) {
+    while(not ready_to_write.isExpired()) modm::this_fiber::yield();
+    while(not this->acquireMaster()) modm::this_fiber::yield();
+    Cs::reset();
+
+    closure();
```

Comment:
```suggestion
    std::forward<Closure>(closure)();
```
#CosCppBullshitAboutPerfectForwarding

Reply:
The lambda capture counts as argument the perfect forwarding is required for, right?

Reply:
aiiieee 🙈